analyzers/src/SonarAnalyzer.CSharp/Rules/PrivateStaticMethodUsedOnlyByNestedClass.cs

```diff
@@ -0,0 +1,190 @@
+ */
+
+using SonarAnalyzer.CFG.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class PrivateStaticMethodUsedOnlyByNestedClass : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S3398";
+    private const string MessageFormat = "Move the method inside '{0}'.";
+
+    private static readonly SyntaxKind[] AnalyzedSyntaxKinds = new[]
+    {
+        SyntaxKind.ClassDeclaration,
+        SyntaxKind.StructDeclaration,
+        SyntaxKind.InterfaceDeclaration,
+        SyntaxKindEx.RecordClassDeclaration,
+        SyntaxKindEx.RecordStructDeclaration
+    };
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+            {
+                var declaredType = (TypeDeclarationSyntax)c.Node;
+
+                if (!IsPartial(declaredType)
+                    && HasNestedTypeDeclarations(declaredType)
+                    && PrivateStaticMethodsOf(declaredType) is { Length: > 0 } candidates)
+                {
+                    var methodReferences = TypesWhichUseTheMethods(candidates, declaredType, c.SemanticModel);
+
+                    foreach (var reference in methodReferences)
+                    {
+                        var typeToMoveInto = LowestCommonAncestorOrSelf(reference.Types);
+                        if (typeToMoveInto != declaredType)
+                        {
+                            var nestedTypeName = typeToMoveInto.Identifier.ValueText;
+                            c.ReportIssue(Diagnostic.Create(Rule, reference.Method.Identifier.GetLocation(), nestedTypeName));
+                        }
+                    }
+                }
+            },
+            AnalyzedSyntaxKinds);
+
+    private static MethodDeclarationSyntax[] PrivateStaticMethodsOf(TypeDeclarationSyntax type) =>
+        type.Members
+                .OfType<MethodDeclarationSyntax>()
+                .Where(x => IsPrivateAndStatic(x, type))
+                .ToArray();
+
+    private static bool HasNestedTypeDeclarations(TypeDeclarationSyntax type) =>
+        type.Members
+                .OfType<TypeDeclarationSyntax>()
+                .Any();
+
+    private static bool IsPrivateAndStatic(MethodDeclarationSyntax method, TypeDeclarationSyntax containingType) =>
+        method.Modifiers.Any(x => x.IsKind(SyntaxKind.StaticKeyword))
+        && ((HasAnyModifier(method, SyntaxKind.PrivateKeyword) && !HasAnyModifier(method, SyntaxKind.ProtectedKeyword))
+           || (!HasAnyModifier(method, SyntaxKind.PublicKeyword, SyntaxKind.ProtectedKeyword, SyntaxKind.InternalKeyword) && IsClassOrRecordClassOrInterfaceDeclaration(containingType)));
```

Comment:
why the difference between
(is Private and not Private Protected)
and
(is not Public, Protected or Internal)

Reply:
The analyzer is only looking for private (static) methods. A method is private if:
- it's explicitly declared as private (and it's not private protected)
- the accessibility is not specified, but it's in a class, record class, or interface, where the default accessibility is private